Fix Field mapping bug - forgets related fields

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Fix Field mapping bug - forgets related fields

I hit this during r-run on imports and initially thought it was a regression but it is there in 5.48 too.

To replicate
1) do an import with a contact where some of the fields are for a related contact - [here is the csv I used for testing](https://drive.google.com/file/d/1Qnxs5uQv6hpBgfJw-hYbIRzU1WMjEoia/view?usp=sharing)
2) do NOT save the mapping, do not collect $200
3) once you hit the preview field go back (still no $200)
4) check if the related contact's field is still populated

Before
----------------------------------------
![image](https://user-images.githubusercontent.com/336308/169634415-3ea7cd8b-7c7d-4ae3-abf5-47ca73b34d03.png)

After
----------------------------------------
![image](https://user-images.githubusercontent.com/336308/169634404-9f829920-10a7-45ba-b6a4-442b73071ce1.png)

Technical Details
----------------------------------------
I think I got to the bottom of what the js was doing & switched to a simple jquery hide (well a whole tonne of them - I'm sure it could be more elegant at the jquery end but figuring out this much was a lot....

Comments
----------------------------------------

eileenmcnaughton committed:
Fix Field mapping bug - forgets related fields

aydun commented:
The first time the mapping page is shown, I get additional empty dropdowns after some fields:
![image](https://user-images.githubusercontent.com/2730045/169963262-365a8661-a0cc-4500-83c8-818dcba45a9a.png)

If I then 'continue' to the preview and then 'previous' to the mapping I correctly get:
![image](https://user-images.githubusercontent.com/2730045/169963487-1db35d66-cb17-49bc-86b2-31a4a8acffa7.png)

I added a phone field to trigger the 4th dropdown. The mapping are preserved when returning to the mapping screen from the preview:
![image](https://user-images.githubusercontent.com/2730045/169964266-0a90b08d-7439-421b-852a-887abfbf64f1.png)

One small anomaly: when selecting 'street address' for the main contact, the default location is 'Primary', but when selecting 'child of', 'street address', the default location is 'Home' instead of 'Primary'.

aydun commented:
I see you've pushed more changes - those results were from before that push.

eileenmcnaughton commented:
@aydun no - just a rebase - there weren't any changes in it - just keeping it fresh

eileenmcnaughton commented:
@aydun I'm assuming your anomaly is unchanged by this PR?

aydun commented on CRM/Contact/Import/Form/MapField.php:
```diff
@@ -595,4 +595,31 @@ private function isLocationTypeRequired($name): bool {
+      foreach ([1, 2, 3] as $columnNumber) {
+        if (empty($mapperValues[$columnNumber])) {
+          $rowNumber = str_replace('mapper[', '', substr($row, 0, -1));
+          $js[] = "cj('#mapper_{$rowNumber}_{$columnNumber}').hide();";
```
Isn't 'cj' deprecated?  How about:
`$js[] = "CRM.\$('#mapper_{$rowNumber}_{$columnNumber}').hide();";`

eileenmcnaughton replied:
yeah - I'm always confused about that - it doesn't seem to work on some pages  - @colemanw what should I use (Or are we just happy with 'a bit better' here?)

aydun commented:
> @aydun I'm assuming your anomaly is unchanged by this PR?
Yes - this also happens without this PR.  I just noticed it while testing this, so not a problem for this PR.

eileenmcnaughton commented:
@aydun I think the 'Primary' behaviour is better - so I might yet circle round to that (not necessarily in this PR)

eileenmcnaughton commented on CRM/Contact/Import/Form/MapField.php:
```diff
@@ -289,16 +288,18 @@ public function buildQuickForm() {
-        $js .= "swapOptions($formName, 'mapper[$i]', 0, 3, 'hs_mapper_0_');\n";
         if ($hasColumnNames) {
           // do array search first to see if has mapped key
           $columnKey = array_search($this->_columnNames[$i], $this->getFieldTitles());
```
@aydun - I can add that in - but do you want to confirm if it works with the other change you suggested? - if so I can include that too

(I think you can actually suggest changes in review & I can commit them through th eUI too)

eileenmcnaughton committed:
Additional defaults fix

aydun committed:
Show secondary fields (eg location for address fields) when setting
default mappings.

Use CRM.$ instead of cj for new jQuery

Fix an assignment to $this->fieldUsed.  Not sure why it exists but
lets update the right key.

aydun commented:
See: https://github.com/eileenmcnaughton/civicrm-core/pull/14

eileenmcnaughton commented:
@aydun - I merged your commit in & rebased to cheer up Jenkins - but now it's defaulting to 'Home' for both rather than 'Primary' for both

aydun commented:
> but now it's defaulting to 'Home' for both rather than 'Primary' for both

That's odd ... it's showing Primary for me on initial display:

![image](https://user-images.githubusercontent.com/2730045/170119704-2b290d6e-0f4c-4680-921d-ef3082f3301d.png)

eileenmcnaughton commented:
I'll try again

eileenmcnaughton commented:
@aydun - OK it IS showing 'Primary' for main contact - but when I changed 2 fields to related I got 'Home'

![image](https://user-images.githubusercontent.com/336308/170121572-1592c261-b1b3-405a-a368-a69b9b488817.png)

eileenmcnaughton committed:
Update test for latest iteration

darrick commented on CRM/Contact/Import/Form/MapField.php:
```diff
@@ -595,4 +594,57 @@ private function isLocationTypeRequired($name): bool {
+    if (!empty($sel2[$key]) && is_array($sel2[$key])) {
+      reset($sel2[$key]);
+      $subkey = key($sel2[$key]);
+      $def[] = $subkey;
```
I think "Primary" is set here when the form is first loaded and the mapping is guesstimated.  But after the form is built there is no default set for $sel2 so "Home" is selected as that is the first item in the select list.

darrick replied:
The options are created via JS here: https://github.com/civicrm/civicrm-packages/blob/96b7c2441b851395176bb275a45e24fbe31e3187/HTML/QuickForm/hierselect.php#L427

eileenmcnaughton replied:
I've just rebased this - do you see a fix for ^^ @darrick - it was kind of a tangent on this PR  - so I'm also happy to drop that bit out of scope for this PR

aydun replied:
> I think "Primary" is set here when the form is first loaded and the mapping is guesstimated. 

Yes,  'Primary' is selected because it is first in the php list.  

When you change to a relationship ('spouse of') then the JS `swapoptions()' function is called.  No value is initially set so the dropdown shows the first option.  

The array ordering in php is:
```
"Primary":"Primary",
"5":"Billing",
"1":"Home",
"3":"Main",
"4":"Other",
"2":"Work"
```
However, although the php values are passed in the same order, when the JS function loops through the array to add the select options, it does so based on the ordering of the keys, not the order added, and so the options are ordered as:

```
1: "Home"
​​2: "Work"
​​3: "Main"
​​4: "Other"
​​5: "Billing"
​​Primary: "Primary"
```

To fix the default, we could explicitly pass through a default value and use that when creating the select options in `swapoptions()`

If we wanted to solve the more general issue of making the JS options match the PHP order, I think we would need to assign new key values when we pass from PHP to JS, but then also map back from from JS keys to PHP keys.  That sounds like a lot of effort for something few people are likely to notice!

This looks like it goes well beyond the scope of the initial aim of this PR.

eileenmcnaughton commented:
@colemanw  - this should be mergeable based on ^^ - ie this is an agreed fix - discussion focusses on a tangential bug that came up in review

darrick commented:
This is a bit out of my skill set.  But most things are.  But I did step through the code a bit the other night.   To me it seemed like setDefault was for setting form defaults when the page is first loaded.   And this bug has more to do with the form state being lost or mangled when hitting the back button. 

Also, the csv file used for testing has duplicate header field.  I recall theleague/cvs throws an error for that.   Maybe that should be caught in Validation earlier.  

I just tested real quick.  If I do the following (i.e. just don't set any defaults if the mapper is set).  Then it works fine except the extra selects not be hidden.  This is based off master.
```
    $mapper = $this->getSubmittedValue('mapper');

    for ($i = 0; $i < $this->_columnCount; $i++) {
      $sel = &$this->addElement('hierselect', "mapper[$i]", ts('Mapper for Field %1', [1 => $i]), NULL);
      $sel->setOptions([$sel1, $sel2, $sel3, $sel4]);

      // Don't set any defaults if we are going to the next page.
      // ... or coming back.
      // But do add the js.
      if (!empty($mapper)) {
        //$sel->setValue($mapper[$i]);
        //$js .= "swapOptions($formName, 'mapper[$i]', 0, 3, 'hs_mapper_0_');\n";
      }
      else if ($this->getSubmittedValue('savedMapping') && $processor->getFieldName($i)) {
        $defaults["mapper[$i]"] = $processor->getSavedQuickformDefaultsForColumn($i);
        $js .= $processor->getQuickFormJSForField($i);
      }
      else {
        $js .= "swapOptions($formName, 'mapper[$i]', 0, 3, 'hs_mapper_0_');\n";
        if ($hasColumnNames) {
          // do array search first to see if has mapped key
          $columnKey = array_search($this->_columnNames[$i], $this->getFieldTitles());
          if (isset($this->_fieldUsed[$columnKey])) {
            $defaults["mapper[$i]"] = $columnKey;
            $this->_fieldUsed[$key] = TRUE;
          }
          else {
            // Infer the default from the column names if we have them
            $defaults["mapper[$i]"] = [
              $this->defaultFromColumnName($this->_columnNames[$i]),
              0,
            ];
          }
        }
        else {
          // Otherwise guess the default from the form of the data
          $defaults["mapper[$i]"] = [
            $this->defaultFromData($this->getDataPatterns(), $i),
            //                     $defaultLocationType->id
            0,
          ];
        }
      }

    }
```
<img width="848" alt="image" src="https://user-images.githubusercontent.com/755624/170801693-bdcda892-9f3d-4d0e-8e2f-e0e4d71cf5bf.png">

darrick commented:
The following solves the problem on my end: https://github.com/civicrm/civicrm-core/commit/c9cd1c113369c6a2fa3b4d60b6cfd14ffd93d99f

Adding this `$js .= "swapOptions($formName, 'mapper[$i]', 0, 3, 'hs_mapper_0_');\n";` is what was causing the second, third and fourth options to get hosed.   As that function rebuilds the options for the select lists above the select list it was called from.   Usually it's called from the onchange event for the appropriate select list.   But above is hard coded to the first select list.

For what @aydun is mentioning about having a default location type maybe the swapOption js could be modified to use a default as it's currently hardcoded to false.  Setting it in buildForm will only work for inferred defaults.
